File: src/main/go/com/lmmmowi/leetcode/go/p90/Solution.go
// 90.子集 II[https://leetcode-cn.com/problems/subsets-ii/]
package main

import "fmt"

func main() {
	result := subsetsWithDup([]int{4, 4, 4, 1, 4})
	for i := range result {
		fmt.Print("LINE==>[")
		for j := range result[i] {
			fmt.Print(result[i][j], ",")
		}
		fmt.Println("]")
	}
}

func subsetsWithDup(nums []int) [][]int {
	result := [][]int{{}}

	sort(nums)
	for n := 1; n <= len(nums); n++ {
		result = find(n, 0, 0, nums, make([]int, len(nums)), result)
	}

	return result
}

func find(n int, m int, k int, nums []int, arr []int, result [][]int) [][]int {
	if n == m {
		item := make([]int, n)
		copy(item, arr)
		return append(result, item)
	}

	first := len(nums) - (n - m) + 1
	if k <= first {
		w := nums[k] + 1
		for i := k; i < first; i++ {
			if nums[i] == w {
				continue
			}
			w = nums[i]

			arr[m] = nums[i]
			result = find(n, m+1, i+1, nums, arr, result)
		}
	}

	return result
}

func sort(arr []int) {
	for i := 0; i < len(arr); i++ {
		for j := i + 1; j < len(arr); j++ {
			if arr[i] > arr[j] {
				temp := arr[i]
				arr[i] = arr[j]
				arr[j] = temp
			}
		}
	}
}

File: src/main/java/com/lmmmowi/leetcode/p90/Solution.java
package com.lmmmowi.leetcode.p90;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;
import java.util.Stack;

/**
 * @Author: lmmmowi
 * @Date: 2020/12/30
 * @Description: 90.子集 II[https://leetcode-cn.com/problems/subsets-ii/]
 */
public class Solution {

    public List<List<Integer>> subsetsWithDup(int[] nums) {
        Arrays.sort(nums);

        List<List<Integer>> results = new ArrayList<>();
        this.find(nums, 0, new Stack<>(), results);
        return results;
    }

    private void find(int[] nums, int index, Stack<Integer> stack, List<List<Integer>> results) {
        results.add(new ArrayList<>(stack));
        if (index >= nums.length) {
            return;
        }

        Integer chosenNum = null;
        for (int i = index; i < nums.length; i++) {
            Integer num = nums[i];
            if (num.equals(chosenNum)) {
                continue;
            }

            chosenNum = num;
            stack.add(num);
            this.find(nums, i + 1, stack, results);
            stack.remove(num);
        }
    }
}
